Add Time#deconstruct_keys

zverok opened the pull request:
I believe that `Time` being suitable for pattern-matching is a reasonable feature with many possible usages, which will increase usability of `Time` and would be a good show case for pattern matching.

**Implementation decisions**

**UPD:** As per [Matz's decision](https://bugs.ruby-lang.org/issues/19071#note-2), only `#deconstruct_keys` is implemented now.

`Time#deconstruct_keys`: 
* chosen keys: `[:year, :month, :day, :yday, :wday, :hour, :min, :sec, :subsec, :dst, :zone]` 
* I am open to discussing whether we should include other subsecond units (or any whatsoever)
* It might be useful (but too loose interface) to support `mon` as a synonym for `month`?.. But might be confusing if somebody will unpack the `**rest`
* `day`, not `mday`, seems most reasonable

Possible usages:
```ruby
case t
in year: ...2022
  puts "too old"
in month: ..9
  puts "quarter 1-3"
in wday: 1..5, month:
  puts "working day in month #{month}"
end

if t in Time(wday: 3, day: ..7)
  puts "first Wednesday of the month"
end
```

~~`Time#to_h`: added on a "why not" basis :) As we already have "convert to hash" in the form of `deconstruct_keys(nil)`, having a canonic form seems harmless. Open for discussion.~~
~~`Time#deconstruct`: returns time components in order `[year, month, mday, hour, min, sec, subsec]`~~

jeremyevans commented on time.c:
```diff
@@ -34,6 +34,7 @@
 #include "id.h"
 #include "internal.h"
 #include "internal/array.h"
+#include "internal/hash.h"
```
@nobu knows more than I do about this, but I think this changes the dependencies, and I don't see an update to the dependency information.

zverok replied:
Can you please elaborate? I think I am missing something big, where's the dependency information that needs to be updated? In Makefiles?..

jeremyevans replied:
The dependencies are stored at the bottom of `common.mk`.  They are autogenerated, I think by `tool/update-deps`.

jeremyevans replied:
Looking at `common.mk`, it definitely needs to be updated.  You can take the line you need to add from the failing CI job: https://github.com/ruby/ruby/actions/runs/3456234830/jobs/5768826516#step:11:13

jeremyevans commented on time.c:
```diff
@@ -4826,6 +4828,102 @@ time_to_a(VALUE time)
+        rb_hash_aset(h, ID2SYM(key_id), val);
+
+#define set_key_if_matches(sym_id, val) \
+        if(ID2SYM(sym_id) == key) rb_hash_aset(h, key, val);
```
I'd prefer to not use macros here, as I think it makes the resulting code more difficult to understand.

zverok replied:
I'll change it if you insist, but honestly, I believe that repeating 11 (eleven) times the construct 
```c
if(ID2SYM(id_year) == key) {
  rb_hash_aset(h, key, tobj->vtm.year);
}
```
...seems to me the way to make it harder to understand, not easier. Currently, it is 100% exposed that we just do the very same thing 11 times.
But with all the `if`s (properly formatted like C codebase prefers) it would be 33 lines which will require some head-scratching to understand whether they are homogenous or not, and whether all of them are just setting the field by condition or there are some more details to follow.

jeremyevans replied:
Certainly the `set_key` macro is not helpful over calling `rb_hash_aset` directly, especially once you switch from static IDs to static symbols.

In terms of `set_key_if_matches`, that is a grayer area.  However, you don't need 33 lines. you can keep the same 11 lines.  Ruby uses `if (cond) do_this;` (single line without braces) all the time (simple grep shows over 4000 uses just in the `*.c` in the root directory).

If you think it must be kept, please move the macro definition to the line before first use, and the undef to the line after last use.  That at least reduces possible confusion.

jeremyevans commented on time.c:
```diff
@@ -48,7 +49,8 @@
 static ID id_local_to_utc, id_utc_to_local, id_find_timezone;
-static ID id_year, id_mon, id_mday, id_hour, id_min, id_sec, id_isdst;
+static ID id_year, id_mon, id_mday, id_hour, id_min, id_sec, id_subsec, id_isdst, id_dst;
+static ID id_month, id_day, id_yday, id_wday;
```
Seems odd to add static IDs for values never used as IDs and only used as symbols.  I recommend using `static VALUE sym_*` for all of the symbols this uses.

zverok replied:
That was done for homogenity. Most of necessary IDs are already defined in the file, so I needed to add only a few, and it was either
1. this (add `ID`s, and use `ID2SYM`), or
2. add symbol representation for those where `ID`s where already present, or
3. break homogeneity (use `ID2SYM` for existing `ID`s, and just symbols for those I added)

Which looks like the most reasonable option to you?..

jeremyevans replied:
I think option 2 is what I recommended, and that sounds like the most reasonable option.

zverok commented:
@jeremyevans You were completely right, macros were overthinking, removed them, and changed IDs to symbols for new keys, it looks cleaner this way.

I still need advice on updating dependencies, `tool/update-deps` didn't change anything :thinking:

jeremyevans commented on time.c:
```diff
@@ -4834,6 +4836,96 @@ time_to_a(VALUE time)
+    if (NIL_P(keys)) {
+        h = rb_hash_new_with_size(11);
+
+        rb_hash_aset(h, ID2SYM(id_year), tobj->vtm.year);
```
After asking my opinion and giving three options, where I said to go with option 2, you implemented option 3. Not sure if that was intentional or not. I suppose it doesn't matter much, both are better than option 1.

zverok commented:
@jeremyevans Yeah, sorry, you are completely right. Trying to "quickly wrap it up" after exhausting week was not a good call of mine :shrug:

jeremyevans reviewed (approved):
Looks good, thanks for working on this!

sampatbadhe commented on time.c:
```diff
@@ -4834,6 +4839,96 @@ time_to_a(VALUE time)
+ *
+ *    t = Time.utc(2022, 10, 5, 21, 25, 30)
+ *
+ *    if t in wday: 3, day: ..7  # uses deconstruct_keys unerneath
```
```suggestion
 *    if t in wday: 3, day: ..7  # uses deconstruct_keys underneath
```

zverok commented:
I updated NEWS and fixed the typo, should I wait for @nobu's review, or can it be merged already? :cat: 
(I have merge rights, just not sure yet of the proper process, the months since I received the rights were ... turbulent)

jeremyevans commented:
Should be fine to merge after CI completes.
